lib/streamaggr/streamaggr.go

```diff
@@ -891,20 +928,28 @@ func (a *aggregator) Push(tss []prompbmarshal.TimeSeries, matchIdxs []byte) {
 		// do not intern key because number of unique keys could be too high
 		key := bytesutil.ToUnsafeString(buf[bufLen:])
 		for _, sample := range ts.Samples {
+			a.muFlushAfter.Lock()
+			a.flushAfter.Update(float64(currentTimestamp - sample.Timestamp))
+			a.muFlushAfter.Unlock()
 			if math.IsNaN(sample.Value) {
 				a.ignoredNanSamples.Inc()
 				// Skip NaN values
+				a.ignoredNanSamples.Inc()
 				continue
 			}
 			if ignoreOldSamples && sample.Timestamp < minTimestamp {
 				a.ignoredOldSamples.Inc()
 				// Skip old samples outside the current aggregation interval
+				a.ignoredOldSamples.Inc()
 				continue
 			}
 			if maxLag < now-sample.Timestamp {
 				maxLag = now - sample.Timestamp
 			}
-			samples = append(samples, pushSample{
+			if ignoreOldSamples {
+				flushIdx = int((sample.Timestamp)/a.tickInterval+1) % aggrStateSize
+			}
+			samples[flushIdx] = append(samples[flushIdx], pushSample{
```

Comment:
As samles is now `[][]pushSample` code a few after this one needs to be updated:
```
	if len(samples) > 0 { // this needs to check for specific flush ID
		a.matchedSamples.Add(len(samples))
		a.samplesLag.Update(float64(maxLag) / 1_000)
	}
```

P.S. left a comment here as github does not allow to leave comments for unchanged lines